File: src/stages/dedupe.js
import { createHash } from "crypto";
import { createReadStream } from "fs";
import { logger } from "../utils/logger.js";
import { progressBar } from "../utils/progress.js";

/**
 * Calculates SHA-256 hashes for all media files in the manifest.
 *
 * This is the first phase of deduplication. It computes a cryptographic hash
 * for each media file to enable content-based duplicate detection. Files with
 * identical hashes are considered duplicates regardless of filename or location.
 *
 * @param {Array<{
 *   mediaPath: string,
 *   json: { path: string, confidence: number } | null,
 *   source: { type: 'album', name: string } | { type: 'loose' }
 * }>} manifest - The manifest of media files to hash.
 * @returns {Promise<void>} Modifies the manifest in-place by adding a `dedupe` property
 *   to each entry containing the calculated hash and initial action state.
 */
export async function dedupeHash(manifest) {
  const bar = progressBar("⧗ Calculating hashes |{bar}| {percentage}% | {value}/{total} entries");
  bar.start(manifest.length, 0);

  for (const [index, entry] of manifest.entries()) {
    try {
      const sha256 = await calculateHash(entry.mediaPath);

      manifest[index] = {
        ...entry,
        dedupe: {
          hash: sha256,
          status: "pending",
        },
      };
    } catch (error) {
      logger.error(error.message);
      manifest[index] = {
        ...entry,
        dedupe: {
          hash: null,
          status: "error",
          reason: error.message,
        },
      };
    }

    bar.increment();
  }

  bar.stop();
}

function calculateHash(filePath) {
  return new Promise((resolve, reject) => {
    const hash = createHash("sha256");
    const stream = createReadStream(filePath);

    stream.on("error", (err) => {
      reject(err);
    });

    stream.on("data", (chunk) => {
      hash.update(chunk);
    });

    stream.on("end", () => {
      resolve(hash.digest("hex"));
    });
  });
}

/**
 * Resolves deduplication actions for files based on content hashes and user preferences.
 *
 * This is the second phase of deduplication. It analyzes files with identical hashes
 * and determines which should be kept or deleted based on the following rules:
 *  - Unique files (no duplicates) are always kept.
 *  - When duplicates have mixed source types (album and loose), files matching the
 *    preferred type are kept and others are deleted.
 *  - When duplicates are all the same type (all album or all loose), all are kept
 *    to preserve files across different albums or locations.
 *
 * @param {Array<{
 *   mediaPath: string,
 *   json: { path: string, confidence: number } | null,
 *   source: { type: 'album', name: string } | { type: 'loose' },
 *   dedupe: { hash: string, status: 'pending' }
 * }>} manifest - The manifest with calculated hashes from dedupeHash.
 * @param {{ prefer: 'album' | 'loose' }} options - Configuration specifying which source
 *   type to prefer when resolving mixed-type duplicates.
 * @returns {Promise<void>} Modifies the manifest in-place by setting each entry's
 *   `dedupe.status` to either 'keep' or 'delete'.
 * @throws {Error} If any entry ends up with an invalid status after resolution.
 */
export async function dedupeResolve(manifest, options) {
  const bar = progressBar("⧗ Resolving duplicates |{bar}| {percentage}% | {value}/{total} entries");

  // Group entries by hash, skipping entries that failed hashing
  const hashes = new Map();
  for (const entry of manifest) {
    const hash = entry.dedupe.hash;
    if (hash) {
      hashes.set(hash, hashes.get(hash) ?? []);
      hashes.get(hash).push(entry);
    }
  }

  bar.start(hashes.size, 0);

  // Resolve duplicates
  const hashGroups = Array.from(hashes.values());
  for (const duplicates of hashGroups) {
    if (duplicates.length === 1) {
      // Unique file - always keep
      duplicates[0].dedupe.status = "keep";
    } else {
      // Multiple files with same hash - check if mixed types
      const hasPreferred = duplicates.some((e) => e.source.type === options.prefer);
      const hasOther = duplicates.some((e) => e.source.type !== options.prefer);

      if (hasPreferred && hasOther) {
        // Mixed types - keep preferred, delete other
        for (const entry of duplicates) {
          entry.dedupe.status = entry.source.type === options.prefer ? "keep" : "delete";
        }
      } else {
        // Same type - keep all
        for (const entry of duplicates) {
          entry.dedupe.status = "keep";
        }
      }
    }

    bar.increment();
  }

  bar.stop();

  checkInvariants(manifest);
}

function checkInvariants(manifest) {
  for (const entry of manifest) {
    const status = entry.dedupe.status;

    if (status !== "keep" && status !== "delete" && status !== "error") {
      throw new Error(
        `Invalid dedupe status "${status}" for file: ${entry.mediaPath}. Expected "keep", "delete", or "error".`
      );
    }
  }
}

File: packages/core/src/pixelkasten/stages/dedupe.py
"""
Content-based deduplication via SHA-256 hashing.

Ported from packages/core/src/stages/dedupe.js.
"""

import hashlib
from collections.abc import Callable

from pixelkasten.manifest import Dedupe, DedupeResult, ManifestEntry, Status
from pixelkasten.configuration import Options


def dedupe_hash(
    manifest: list[ManifestEntry],
    on_progress: Callable[[int], None] | None = None,
) -> None:
    """
    Compute SHA-256 for each manifest entry.

    Mutates entries in-place, setting entry.dedupe with hash and PENDING status.
    On error, sets entry.dedupe with ERROR status.
    """
    for i, entry in enumerate(manifest):
        try:
            sha256 = _calculate_hash(entry.media_path)
            entry.dedupe = Dedupe(status=Status.PENDING, hash=sha256)
        except Exception as e:
            entry.dedupe = Dedupe(status=Status.ERROR, error=str(e))
        if on_progress is not None:
            on_progress(i + 1)


def dedupe_resolve(manifest: list[ManifestEntry], options: Options) -> None:
    """
    Resolve duplicates by mode.

    Takeout mode:
      - Unique files: KEEP.
      - Mixed source types: keep the side that matches --prefer, delete others.
      - All same source type: keep all.
    Archive mode:
      - Unique files: KEEP.
      - Multiple: keep the lex-smallest media_path, delete the rest.

    Mutates entries in-place, setting dedupe.status and dedupe.result.
    """
    # Group by hash, skipping errored entries.
    groups: dict[str, list[tuple[ManifestEntry, Dedupe]]] = {}
    for entry in manifest:
        if entry.dedupe is None or entry.dedupe.hash is None:
            continue
        groups.setdefault(entry.dedupe.hash, []).append((entry, entry.dedupe))

    if options.mode == "archive":
        _resolve_archive(groups)
    else:
        _resolve_takeout(groups, options.prefer)

    _check_invariants(manifest)


def _resolve_takeout(
    groups: dict[str, list[tuple[ManifestEntry, Dedupe]]],
    prefer: str,
) -> None:
    for duplicates in groups.values():
        if len(duplicates) == 1:
            _, dedupe = duplicates[0]
            dedupe.status = Status.PROCESSED
            dedupe.result = DedupeResult.KEEP
            continue

        has_preferred = any(e.source.type == prefer for e, _ in duplicates)
        has_other = any(e.source.type != prefer for e, _ in duplicates)

        for entry, dedupe in duplicates:
            dedupe.status = Status.PROCESSED
            if has_preferred and has_other:
                dedupe.result = (
                    DedupeResult.KEEP if entry.source.type == prefer else DedupeResult.DELETE
                )
            else:
                dedupe.result = DedupeResult.KEEP


def _resolve_archive(groups: dict[str, list[tuple[ManifestEntry, Dedupe]]]) -> None:
    for duplicates in groups.values():
        if len(duplicates) == 1:
            _, dedupe = duplicates[0]
            dedupe.status = Status.PROCESSED
            dedupe.result = DedupeResult.KEEP
            continue

        winner_path = min(entry.media_path for entry, _ in duplicates)
        for entry, dedupe in duplicates:
            dedupe.status = Status.PROCESSED
            dedupe.result = (
                DedupeResult.KEEP if entry.media_path == winner_path else DedupeResult.DELETE
            )


def _calculate_hash(file_path: str) -> str:
    """Read file in chunks and return hex SHA-256 digest."""
    h = hashlib.sha256()
    with open(file_path, "rb") as f:
        while True:
            chunk = f.read(8192)
            if not chunk:
                break
            h.update(chunk)
    return h.hexdigest()


def _check_invariants(manifest: list[ManifestEntry]) -> None:
    """Verify all dedupe entries have a valid result or error status."""
    for entry in manifest:
        if entry.dedupe is None:
            continue
        if entry.dedupe.status == Status.ERROR:
            continue
        if entry.dedupe.result is None:
            raise ValueError(f"Dedupe result not set for {entry.media_path}")
